### PR TITLE
Use SafeChown rather then chown for volumes on NFS

### PR DESCRIPTION
**Pull request by rhatdan**
NFS Servers will thrown ENOTSUPP error if you attempt to
chown a directory to the same UID and GID as the directory
already has. If volumes are stored on NFS directories this
throws an ugly error and then works on the next try.

Bottom line don't chown directories that already have the correct
UID and GID.

Fixes: https://github.com/containers/podman/issues/14766

[NO NEW TESTS NEEDED] Difficult to setup an NFS Server in testing.

Signed-off-by: Daniel J Walsh <dwalsh@redhat.com>

#### Does this PR introduce a user-facing change?

```release-note
None

```

### CODE REVIEW

**Comment by rhatdan** [comment]
@[berndbausch](https://github.com/berndbausch)
berndbausch could you try this out and see if it fixes your issue

**Comment by berndbausch** [comment]
> @[berndbausch](https://github.com/berndbausch) could you try this out and see if it fixes your issue

TLDR: No, sorry.

I am not a developer and not familiar with the process how to try this out. Having said that, I did the following: Clone Podman version 4.0.3 on my Fedora VM, install the Go compiler and related packages as per the Podman installation instructions, add/change the two lines in libpod/runtime_volume_linux.go, compile Podman (`make BUILDTAGS="selinux seccomp systemd"`). Then use the compiled binary `bin/podman` to create the volume and run the container. 

Unfortunately, same error. I can't tell if I did something wrong or your modification is not effective here. The error message in the code, by the way, is "chowning volume directory %q to %d:%d", which is not the message I get (but perhaps the "chowning" message is written somewhere else, and I definitely don't know how the code is structured).

**Comment by rhatdan** [comment]
@What is the actual error you are seeing?  I updated the PR with a Chown I had missed.

**Comment by berndbausch** [comment]
> @what is the actual error you are seeing? I updated the PR with a Chown I had missed.

Sorry for the delay. It doesn't help that I am on the other side of the planet from the US East Coast. The error I am getting is 
```
Error: lchown /var/lib/containers/storage/volumes/mynfs/_data: operation not permitted
```
Where *mynfs* is the volume's name. This error may come from `vendor/github.com/containers/storage/pkg/system/lchown.go`.

I should have made that more prominent in the issue description.

**Comment by rhatdan** [comment]
I think this would mean that the storage is owned by a different user then the user of podman?  
What is the podman command you are using?
Could you show the output of 
id 
and 
ls -l /var/lib/containers/storage/volumes/mynfs

This indicates to me that mynfs is not owned by the root user?  if you are running Podman as root?  Or are you running in rootless mode?

**Comment by berndbausch** [comment]
It's root all around. Rootless containers can't deal with NFS-based volumes, since the NFS filesystem can be mounted in a user namespace.

I found, meanwhile, that setting the `no_root_squash` option on the NFS server makes the problem go away. Which makes sense, since this option means that root on the NFS client is root on the NFS server and can therefore perform the lchown.

List of my commands, plus the mountpoint permissions:
```
$ sudo podman volume create mynfs --driver local --opt type=nfs --opt o=rw --opt device=192.168.1.16:/srv/nfs

$ sudo ls -lZ /var/lib/containers/storage/volumes/mynfs
total 0
drwxr-xr-x. 2 root root system_u:object_r:container_file_t:s0 6 Jul  1 21:20 _data

sudo podman run -it --rm -v mynfs:/myvol fedora sh
Error: lchown /var/lib/containers/storage/volumes/mynfs/_data: operation not permitted
```
It's also not due to SELinux (or Apparmor, on the Ubuntu system where I did the same thing); I set it to permissive to ensure it doesn't interfere.

**Comment by rhatdan** [comment]
The weird thing is the safechown should see that the file is owned by root and not attampt the chown, with this patch.  But you are seeing it still attempting the chown.

**Comment by berndbausch** [comment]
As far as I am concerned, personally, I don't need this problem to be pursued at all costs. It looks like a niche problem not many people would have, and in my case, I am only experimenting with Podman in order to develop a training course. I would be very happy to contribute to its solution, but don't need the Podman lead developer to spend all his time on it.

So if you are fine, I am happy to close this issue.

**Comment by rhatdan** [comment]
Well this does not hurt anything and should fix problems with NFS backends for volume storage.  I think we should add it.

@containers/podman-maintainers PTAL

**Review by saschagrunert — APPROVED** [review]
LGTM, CI seems down, though.

**Comment by giuseppe** [comment]
> Well this does not hurt anything and should fix problems with NFS backends for volume storage. I think we should add it.

it adds a `Stat` for each `Chown`, which is very expensive on NFS.  I think we should first fix `SafeChown` to attempt the chown and then do the stat on errors.

**Comment by giuseppe** [comment]
> it adds a `Stat` for each `Chown`, which is very expensive on NFS. I think we should first fix `SafeChown` to attempt the chown and then do the stat on errors.

opened a PR for c/storage to address this problem: https://github.com/containers/storage/pull/1281

**Comment by rhatdan** [comment]
@containers/podman-maintainers PTAL
@vrothberg @giuseppe @flouthoc @mheon @baude @ashley-cui PTAL

**Review by flouthoc — APPROVED** [review]
LGTM

**Review by vrothberg — COMMENTED** [review]
/lgtm
